### DISCUSSION TITLE
Help unravel a tangle of dependencies

### DISCUSSION THREAD

**Discussion by vasigorc** [open-telemetry/opentelemetry-java#4229] [Q&A]
Hello everyone 👋 

As we continue adopting Opentelemetry Java SDK in our Scala application we ran across a new puzzle. 

We have this code block:

```scala
  private def recordGaugeMetric(unit: String, metric: OurMetric, value: Double): Unit = {
    meter
      .gaugeBuilder(metric.metricName.value)
      .setUnit(unit)
      .buildWithCallback(measurement => {
        measurement.record(value, toOtelAttributes(metric.labels))
      })
  }
```

and when we execute this code we get `NoSuchMethodError` for `buildWithCallback`:

```shell
Uncaught error from thread [our-app-test-akka.actor.default-dispatcher-7]: 'io.opentelemetry.api.metrics.ObservableDoubleGauge io.opentelemetry.api.metrics.DoubleGaugeBuilder.buildWithCallback(java.util.function.Consumer)', shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[our-app-test]
Uncaught error from thread [our-app-test-akka.actor.default-dispatcher-6]: 'io.opentelemetry.api.metrics.ObservableDoubleGauge io.opentelemetry.api.metrics.DoubleGaugeBuilder.buildWithCallback(java.util.function.Consumer)', shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[our-app-test]
java.lang.NoSuchMethodError: 'io.opentelemetry.api.metrics.ObservableDoubleGauge io.opentelemetry.api.metrics.DoubleGaugeBuilder.buildWithCallback(java.util.function.Consumer)'
	at com.our.app.actors.observability.OtelMetricsConsumer.io$our$app$actors$observability$OtelMetricsConsumer$$recordGaugeMetric(OtelMetricsConsumer.scala:39)
	at com.our.app.actors.observability.OtelMetricsConsumer$$anonfun$wrappedReceive$1.applyOrElse(OtelMetricsConsumer.scala:32)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at io.our.app.actors.mixins.ActorStack$$anonfun$receive$1.applyOrElse(ActorStack.scala:22)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at io.our.app.actors.MdcActorLogging$$anonfun$receive$1.applyOrElse(MdcActorLogging.scala:47)

```

the related dependencies are declared (using `sbt`) like this:

```scala
  lazy val otelVersion = "1.11.0"
  lazy val otlpMetricsVersion = "1.11.0-alpha"
...
    "io.opentelemetry" % "opentelemetry-api" % otelVersion,
    "io.opentelemetry" % "opentelemetry-exporter-otlp" % otelVersion,
    "io.opentelemetry" % "opentelemetry-exporter-otlp-metrics" % otlpMetricsVersion,
    "io.opentelemetry" % "opentelemetry-extension-annotations" % otelVersion,
    "io.opentelemetry" % "opentelemetry-sdk" % otelVersion,
    "io.opentelemetry" % "opentelemetry-sdk-metrics" % otlpMetricsVersion,
    "io.opentelemetry.instrumentation" % "opentelemetry-logback-1.0" % "1.9.2-alpha",
```

Dependency tree shows that everything but `openteletry-api-metrics` (transient dependency for `opentelemetry-logback-1.0`) is from `1.11.0`. In IntelliJ when I click on `buildWithCallback` it takes me to `DoubleGaugeBuilder` from `opentelemetry-api-metrics.-1.9.1.jar`. Please advise 🤔 
<img width="952" alt="Screen Shot 2022-03-02 at 3 05 21 PM" src="https://user-images.githubusercontent.com/7588095/156441407-405c4863-0f21-4663-81a2-4795f95735f4.png">

**Comment by jkwatson** [comment]
You need to update your `io.opentelemetry.instrumentation:opentelemetry-logback-1.0` to the matching version of the API and SDK. It's pulling in an old version.

**Reply by vasigorc** [reply]
`opentelemetry-logback-1.0` is already at it's maximum version: https://search.maven.org/artifact/io.opentelemetry.instrumentation/opentelemetry-logback-1.0

Moreover `opentelemetry-api-metrics` is almost at it's maximum version as the result of the above: https://search.maven.org/artifact/io.opentelemetry/opentelemetry-api-metrics

**Reply by anuraaga** [reply]
Hi @vasigorc - sorry for the churn while working on our alpha modules. The log artifacts were renamed recently

https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v1.10.0

**Reply by vasigorc** [reply]
Thanks @anuraaga ! That was very helpful 👍 

<img width="789" alt="Screen Shot 2022-03-03 at 9 38 09 AM" src="https://user-images.githubusercontent.com/7588095/156586767-9b4ee812-635f-471a-a9b7-af81175a3eb5.png">
